ncurses-6.3/man/clear.1
.\"***************************************************************************
.\" Copyright 2018,2020 Thomas E. Dickey                                     *
.\" Copyright 1998-2016,2017 Free Software Foundation, Inc.                  *
.\"                                                                          *
.\" Permission is hereby granted, free of charge, to any person obtaining a  *
.\" copy of this software and associated documentation files (the            *
.\" "Software"), to deal in the Software without restriction, including      *
.\" without limitation the rights to use, copy, modify, merge, publish,      *
.\" distribute, distribute with modifications, sublicense, and/or sell       *
.\" copies of the Software, and to permit persons to whom the Software is    *
.\" furnished to do so, subject to the following conditions:                 *
.\"                                                                          *
.\" The above copyright notice and this permission notice shall be included  *
.\" in all copies or substantial portions of the Software.                   *
.\"                                                                          *
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS  *
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF               *
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.   *
.\" IN NO EVENT SHALL THE ABOVE COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,   *
.\" DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR    *
.\" OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR    *
.\" THE USE OR OTHER DEALINGS IN THE SOFTWARE.                               *
.\"                                                                          *
.\" Except as contained in this notice, the name(s) of the above copyright   *
.\" holders shall not be used in advertising or otherwise to promote the     *
.\" sale, use or other dealings in this Software without prior written       *
.\" authorization.                                                           *
.\"***************************************************************************
.\"
.\" $Id: clear.1,v 1.23 2020/02/02 23:34:34 tom Exp $
.TH @CLEAR@ 1 ""
.\" these would be fallbacks for DS/DE,
.\" but groff changed the meaning of the macros.
.de NS
.ie n  .sp
.el    .sp .5
.ie n  .in +4
.el    .in +2
.nf
.ft C			\" Courier
..
.de NE
.fi
.ft R
.ie n  .in -4
.el    .in -2
..
.ie \n(.g .ds `` \(lq
.el       .ds `` ``
.ie \n(.g .ds '' \(rq
.el       .ds '' ''
.de bP
.ie n  .IP \(bu 4
.el    .IP \(bu 2
..
.ds n 5
.SH NAME
\fB@CLEAR@\fR \- clear the terminal screen
.SH SYNOPSIS
\fB@CLEAR@\fR [\fB\-T\fR\fItype\fR] [\fB\-V\fP] [\fB\-x\fP]
.br
.SH DESCRIPTION
\fB@CLEAR@\fR clears your screen if this is possible,
including its scrollback buffer
(if the extended \*(``E3\*('' capability is defined).
\fB@CLEAR@\fR looks in the environment for the terminal type
given by the environment variable \fBTERM\fP,
and then in the
\fBterminfo\fR database to determine how to clear the screen.
.PP
\fB@CLEAR@\fR writes to the standard output.
You can redirect the standard output to a file (which prevents
\fB@CLEAR@\fR from actually clearing the screen),
and later \fBcat\fP the file to the screen, clearing it at that point.
.SH OPTIONS
.PP
.TP 5
.B \-T \fItype\fP
indicates the \fItype\fR of terminal.
Normally this option is
unnecessary, because the default is taken from the environment
variable \fBTERM\fR.
If \fB\-T\fR is specified, then the shell
variables \fBLINES\fR and \fBCOLUMNS\fR will also be ignored.
.TP
.B \-V
reports the version of ncurses which was used in this program, and exits.
The options are as follows:
.TP
.B \-x
do not attempt to clear the terminal's scrollback buffer
using the extended \*(``E3\*('' capability.
.SH HISTORY
A \fBclear\fP command appeared in 2.79BSD dated February 24, 1979.
Later that was provided in Unix 8th edition (1985).
.PP
AT&T adapted a different BSD program (\fBtset\fP) to make
a new command (\fBtput\fP),
and used this to replace the \fBclear\fP command with a shell script
which calls \fBtput clear\fP, e.g.,
.NS
/usr/bin/tput ${1:+-T$1} clear 2> /dev/null
exit
.NE
.PP
In 1989, when Keith Bostic revised the BSD \fBtput\fP command
to make it similar to the AT&T \fBtput\fP,
he added a shell script for the \fBclear\fP command:
.NS
exec tput clear
.NE
.PP
The remainder of the script in each case is a copyright notice.
.PP
The ncurses \fBclear\fP command began in 1995 by adapting the original
BSD \fBclear\fP command (with terminfo, of course).
.PP
The \fBE3\fP extension came later:
.bP
In June 1999, xterm provided an extension to the standard control
sequence for clearing the screen.
Rather than clearing just the visible part of the screen using
.NS
printf '\\033[2J'
.NE
.IP
one could clear the \fIscrollback\fP using
.NS
printf '\\033[\fB3\fPJ'
.NE
.IP
This is documented in \fIXTerm Control Sequences\fP as a feature originating
with xterm.
.bP
A few other terminal developers adopted the feature, e.g., PuTTY in 2006.
.bP
In April 2011, a Red Hat developer submitted a patch to the Linux
kernel, modifying its console driver to do the same thing.
The Linux change, part of the 3.0 release, did not mention xterm,
although it was cited in the Red Hat bug report (#683733)
which led to the change.
.bP
Again, a few other terminal developers adopted the feature.
But the
next relevant step was a change to the \fBclear\fP program in 2013
to incorporate this extension.
.bP
In 2013, the \fBE3\fP extension was overlooked in \fB@TPUT@\fP with
the \*(``clear\*('' parameter.
That was addressed in 2016 by reorganizing \fB@TPUT@\fP to share
its logic with \fB@CLEAR@\fP and \fB@TSET@\fP.
.SH PORTABILITY
Neither IEEE Std 1003.1/The Open  Group  Base  Specifications  Issue  7
(POSIX.1-2008) nor X/Open Curses Issue 7 documents @TSET@ or @RESET@.
.PP
The latter documents \fBtput\fP, which could be used to replace this utility
either via a shell script or by an alias (such as a symbolic link) to
run \fB@TPUT@\fP as \fB@CLEAR@\fP.
.SH SEE ALSO
\fB@TPUT@\fR(1), \fBterminfo\fR(\*n)
.PP
This describes \fBncurses\fR
version @NCURSES_MAJOR@.@NCURSES_MINOR@ (patch @NCURSES_PATCH@).


ncurses-6.3/NEWS
-------------------------------------------------------------------------------
-- Copyright 2018-2020,2021 Thomas E. Dickey                                 --
-- Copyright 1998-2017,2018 Free Software Foundation, Inc.                   --
--                                                                           --
-- Permission is hereby granted, free of charge, to any person obtaining a   --
-- copy of this software and associated documentation files (the             --
-- "Software"), to deal in the Software without restriction, including       --
-- without limitation the rights to use, copy, modify, merge, publish,       --
-- distribute, distribute with modifications, sublicense, and/or sell copies --
-- of the Software, and to permit persons to whom the Software is furnished  --
-- to do so, subject to the following conditions:                            --
--                                                                           --
-- The above copyright notice and this permission notice shall be included   --
-- in all copies or substantial portions of the Software.                    --
--                                                                           --
-- THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS   --
-- OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF                --
-- MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN --
-- NO EVENT SHALL THE ABOVE COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,       --
-- DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR     --
-- OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE --
-- USE OR OTHER DEALINGS IN THE SOFTWARE.                                    --
--                                                                           --
-- Except as contained in this notice, the name(s) of the above copyright    --
-- holders shall not be used in advertising or otherwise to promote the      --
-- sale, use or other dealings in this Software without prior written        --
-- authorization.                                                            --
-------------------------------------------------------------------------------
-- $Id: NEWS,v 1.3732 2021/10/20 22:49:51 tom Exp $
-------------------------------------------------------------------------------

This is a log of changes that ncurses has gone through since Zeyd started
working with Pavel Curtis' original work, pcurses, in 1992.

Changes through 1.9.9e are recorded by Zeyd M Ben-Halim.
Changes since 1.9.9e are recorded by Thomas E Dickey.

Contributors include those who have provided patches (even small ones), as well
as those who provide useful information (bug reports, analyses).  Changes with
no cited author are the work of Thomas E Dickey (TD).

A few contributors may be cited in this file by their initials.
Each accounts for half of one percent or more of the changes since 1.9.9e.
See the AUTHORS file for the corresponding full names.

Changes through 1.9.9e did not credit all contributions;
it is not possible to add this information.

20211021 6.3 release for upload to ftp.gnu.org
	+ update release notes
	+ add "ncu2openbsd" script, to illustrate how to update an OpenBSD
	  system to use a current ncurses release.

20211018
	+ check for screen size-change in scr_init() and scr_restore(), in case
	  a screen dump does not match the current screen dimensions (report by
	  Frank Tkalcevic).

20211017
	+ amend change for pkg-config to account for "none" being returned in
	  the libdir-path result rather than "no" (report by Gabriele Balducci).

20211016
	+ build-fix for pmake with libtool.
	+ improve make-tar.sh scripts, adding COPYING to tar file, and clean up
	  shellcheck warnings.
	+ add link for "reset6" manpage in test-package ncurses6-doc
	+ revise configure option --with-pkg-config-libdir, using the actual
	  search path from pkg-config or pkgconf using the output from --debug
	  (report by Pascal Pignard).
	+ freeze ABI in ".map" files.

20211009
	+ implement "+m" option in tabs program.
	+ fill in some details for infoton -TD
	+ fix spelling/consistency in several descriptions -TD
	+ use vt420+lrmm in vt420 -TD
	+ modify save_tty_settings() to avoid opening /dev/tty for cases other
	  than reset/init, e.g., for clear.
	+ modify output of "toe -as" to show first description found rather
	  than the last.
	+ improve tic checks for number of parameters of smglp, smgrp, smgtp,
	  and smgbp (cf: 20020525).
	+ correct off-by-one comparison in last_char(), which did not allow
	  special case of ":" in a terminfo description field (cf: 20120407).
	+ remove check in tic that assumes that none or both parameterized and
	  non-parameterized margin-setting capabilities are present
	  (cf: 20101002).

20211002
	+ use return-value from vsnprintf to reallocate as needed to allow for
	  buffers larger than the screen size (report by "_RuRo_").
	+ modify tset "-q" option to refrain from modifying terminal modes, to
	  match the documentation.
	+ add section on margins to terminfo.5, adapted from X/Open Curses.
	+ make tput/tset warning messages consistently using alias names when
	  those are used, rather than the underlying program's name.
	+ improve tput usage message for aliases such as clear, by eliminating
	  tput-specific portions.
	+ add a check in toe to ensure that a "termcap file" is text rather
	  than binary.
	+ further build-fixes for OpenBSD 6.9, whose header files differ from
	  the other BSDs.

20210925
	+ add kbeg to xterm+keypad to accommodate termcap applications -TD
	+ add smglp and smgrp to vt420+lrmm, to provide useful data for the
	  "tabs" +m option -TD
	+ build-fix for gcc 3.4.3 with Solaris10, which does not allow forward
	  reference of anonymous struct typedef.
	+ modify tput to allow multiple commands per line.
	+ minor fixes for tset manpage.

20210911
	+ adjust ifdef in test_opaque.c to fix build with ncurses 5.7

ncurses-6.3/progs/clear.c
/****************************************************************************
 * Copyright 2020,2021 Thomas E. Dickey                                     *
 * Copyright 1998-2016,2017 Free Software Foundation, Inc.                  *
 *                                                                          *
 * Permission is hereby granted, free of charge, to any person obtaining a  *
 * copy of this software and associated documentation files (the            *
 * "Software"), to deal in the Software without restriction, including      *
 * without limitation the rights to use, copy, modify, merge, publish,      *
 * distribute, distribute with modifications, sublicense, and/or sell       *
 * copies of the Software, and to permit persons to whom the Software is    *
 * furnished to do so, subject to the following conditions:                 *
 *                                                                          *
 * The above copyright notice and this permission notice shall be included  *
 * in all copies or substantial portions of the Software.                   *
 *                                                                          *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS  *
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF               *
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.   *
 * IN NO EVENT SHALL THE ABOVE COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,   *
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR    *
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR    *
 * THE USE OR OTHER DEALINGS IN THE SOFTWARE.                               *
 *                                                                          *
 * Except as contained in this notice, the name(s) of the above copyright   *
 * holders shall not be used in advertising or otherwise to promote the     *
 * sale, use or other dealings in this Software without prior written       *
 * authorization.                                                           *
 ****************************************************************************/

/****************************************************************************
 *  Author: Zeyd M. Ben-Halim <zmbenhal@netcom.com> 1992,1995               *
 *     and: Eric S. Raymond <esr@snark.thyrsus.com>                         *
 *     and: Thomas E. Dickey                        1996-on                 *
 ****************************************************************************/

/*
 * clear.c --  clears the terminal's screen
 */

#define USE_LIBTINFO
#include <clear_cmd.h>
#include <tty_settings.h>

MODULE_ID("$Id: clear.c,v 1.24 2021/03/20 18:23:14 tom Exp $")

const char *_nc_progname = "clear";

static GCC_NORETURN void
usage(void)
{
#define KEEP(s) s "\n"
    static const char msg[] =
    {
	KEEP("")
	KEEP("Options:")
	KEEP("  -T TERM     use this instead of $TERM")
	KEEP("  -V          print curses-version")
	KEEP("  -x          do not try to clear scrollback")
    };
#undef KEEP
    (void) fprintf(stderr, "Usage: %s [options]\n", _nc_progname);
    fputs(msg, stderr);
    ExitProgram(EXIT_FAILURE);
}

int
main(
	int argc GCC_UNUSED,
	char *argv[]GCC_UNUSED)
{
    TTY tty_settings;
    int fd;
    int c;
    char *term;
    bool opt_x = FALSE;		/* clear scrollback if possible */

    _nc_progname = _nc_rootname(argv[0]);
    term = getenv("TERM");

    while ((c = getopt(argc, argv, "T:Vx")) != -1) {
	switch (c) {
	case 'T':
	    use_env(FALSE);
	    use_tioctl(TRUE);
	    term = optarg;
	    break;
	case 'V':
	    puts(curses_version());
	    ExitProgram(EXIT_SUCCESS);
	case 'x':		/* do not try to clear scrollback */
	    opt_x = TRUE;
	    break;
	default:
	    usage();
	    /* NOTREACHED */
	}
    }
    if (optind < argc)
	usage();

    fd = save_tty_settings(&tty_settings, FALSE);

    setupterm(term, fd, (int *) 0);

    ExitProgram((clear_cmd(opt_x) == ERR)
		? EXIT_FAILURE
		: EXIT_SUCCESS);
}
